utils/src/main/kotlin/OrtAuthenticator.kt

```diff
@@ -127,20 +127,29 @@ fun getNetrcAuthentication(contents: String, machine: String): PasswordAuthentic
 
     val iterator = lines.joinToString(" ").split(Regex("\\s+")).iterator()
 
-    var machineFound = false
     var login: String? = null
     var password: String? = null
+    var lastEntry: String? = null
+    val credentialsPerMachine = mutableMapOf<String, Pair<String, String>>()
 
     while (iterator.hasNext()) {
         when (iterator.next()) {
-            "machine" -> machineFound = iterator.hasNext() && iterator.next() == machine
-            "login" -> login = if (machineFound && iterator.hasNext()) iterator.next() else null
-            "password" -> password = if (machineFound && iterator.hasNext()) iterator.next() else null
-            "default" -> machineFound = true
+            "machine" -> lastEntry = iterator.next()
+            "login" -> login = iterator.next()
+            "password" -> password = iterator.next()
+            "default" -> lastEntry = "default"
         }
 
-        if (login != null && password != null) return PasswordAuthentication(login, password.toCharArray())
+        if (login != null && password != null && lastEntry != null) {
+            credentialsPerMachine[lastEntry] = login to password
+        }
     }
 
+    credentialsPerMachine[machine]?.let {
+        return PasswordAuthentication(it.first, it.second.toCharArray())
+    }
+    credentialsPerMachine["default"]?.let {
+        return PasswordAuthentication(it.first, it.second.toCharArray())
+    }
```

Comment:
How about this instead:

    return listOfNotNull(credentialsPerMachine[machine], credentialsPerMachine["default"]).firstOrNull()
        ?.let { PasswordAuthentication(it.first, it.second.toCharArray()) }

Reply:
The solution before was doing only one map lookup in the best of case and yours is always doing two.
I know we are talking about micro optimizations here but nevertheless, we should not do useless things.

Reply:
how about
```
credentialsPerMachine[machine]?.let {
        return PasswordAuthentication(it.first, it.second.toCharArray())
}
return credentialsPerMachine["default"]?.let {
        PasswordAuthentication(it.first, it.second.toCharArray())
}
```

Reply:
Well, my main aim was to reduce the total amount of code, and in particular to not have the duplication to construct `PasswordAuthentication`.

> The solution before was doing only one map lookup in the best of case and yours is always doing two.
I know we are talking about micro optimizations here but nevertheless, we should not do useless things.

I basically like that line of thinking 😀 But that could be easily fixed by

    return sequenceOf(credentialsPerMachine[machine], credentialsPerMachine["default"]).find { it != null }
        ?.let { PasswordAuthentication(it.first, it.second.toCharArray()) }

Reply:
But is it not the same ?
When writing `sequenceOf(credentialsPerMachine[machine], credentialsPerMachine["default"])` he always do two lookups first even if later, only one is consumed by `find`.

Reply:
Ok I found the solution
```
return (credentialsPerMachine[machine] ?: credentialsPerMachine["default"])?.let {
        PasswordAuthentication(it.first, it.second.toCharArray())
    }
```
but then you have no specific logging. Please advise.

Reply:
No, that's exactly the difference between a sequence and a list: Sequence members are evaluated lazily. So if the first entry already is non-null, the second entry will not even be looked at. Also see e.g. https://www.baeldung.com/kotlin/sequences.

Reply:
> but then you have no specific logging. Please advise.

I made up my mind. let's not over-complicate things by using sequences for only two elements here, but basically use your code as-is. I just have one more formatting nit (in a separate comment).

Reply:
I think there is a misunderstanding here : yes, because this is a sequence, the .find is going to be called only once.
However, because of Kotlin eager evaluation, when calling sequenceOf, it is going to evaluate both two parameters and therefore going to do two lookups.
https://ahsensaeed.com/kotlin-lazy-evaluation-haskell-laziness/

I don't see any lazy in the parameters of `sequenceOf`so I assume all the parameters are evaluated when `sequenceof` is called.

Reply:
You're right in that my usage of `sequenceOf()` was wrong, I should have used `sequence {}` as in [this example](https://pl.kotl.in/dPdSKTM1t).